### PR TITLE
FM2-222: Add create, update, and delete methods for Practitioner

### PR DESCRIPTION
**Pull request by gitcliff**
link:https://issues.openmrs.org/browse/FM2-222

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java [hunk 1]
```diff
@@ -55,6 +59,38 @@ public Practitioner getPractitionerById(@IdParam @NotNull IdType id) {
...
 	}
 	
+
+	@Create
+	public MethodOutcome createPractitioner(@ResourceParam Practitioner practitioner) {
+		return FhirProviderUtils.buildCreate(practitionerService.create(Practitioner30_40.convertPractitioner(practitioner)));
+	} 
+	
+	@Update
+	@SuppressWarnings("unused")
```
Comment (varung-31) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java:69]: Why is there a need to use this annotation ? You have anyways used these methods in your unit tests. Please find this useful throughout the PR.

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java [hunk 2]
```diff
@@ -55,6 +59,38 @@ public Practitioner getPractitionerById(@IdParam @NotNull IdType id) {
...
+	public OperationOutcome deletePractitioner(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Practitioner practitioner = practitionerService.delete(id.getIdPart());
+		if (practitioner == null) {
+			throw new ResourceNotFoundException("Could not find practitioner to delete with id " + id.getIdPart());
+		}
+		OperationOutcome retVal = new OperationOutcome();
+		retVal.setId(id.getIdPart());
+		retVal.getText().setDivAsString("Deleted successfully");
+		return retVal;
```
Comment (corneliouzbett) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java:91]: Make use of util from `FhirProviderUtils.buildDelete` introduced here #214

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java [hunk 3]
```diff
@@ -330,4 +344,93 @@ private void verifyUri(String uri) throws Exception {
 		assertThat(results.getEntry().get(0).getResource(), notNullValue());
 		assertThat(results.getEntry().get(0).getResource().getIdElement().getIdPart(), equalTo(PRACTITIONER_UUID));
 	}
+	
+	@Test
+	public void createPractitioner_shouldCreatePractitioner() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_CREATE_PRACTITIONER_PATH)) {
+			jsonPractitioner = IOUtils.toString(is);
+		}
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java:353]:
```suggestion
		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_CREATE_PRACTITIONER_PATH)) {
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
		}
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java [hunk 4]
```diff
@@ -330,4 +344,93 @@ private void verifyUri(String uri) throws Exception {
...
+		        .accept(FhirMediaTypes.JSON).go();
+
+		assertThat(response, isCreated());
+	}
+
+	@Test
+	public void updatePractitioner_shouldUpdateExistingPractitioner() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_PRACTITIONER_PATH )) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java:370]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java [hunk 5]
```diff
@@ -330,4 +344,93 @@ private void verifyUri(String uri) throws Exception {
...
+				.accept(FhirMediaTypes.JSON).go();
+
+		assertThat(response, isOk());
+	}
+
+	@Test
+	public void updatePractitioner_shouldErrorForNoId() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_PRACTITIONER_NO_ID_PATH)) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java:388]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java [hunk 6]
```diff
@@ -330,4 +344,93 @@ private void verifyUri(String uri) throws Exception {
...
+		assertThat(response.getContentAsString(),
+				containsStringIgnoringCase("body must contain an ID element for update"));
+	}
+
+	@Test
+	public void updatePractitioner_shouldErrorForIdMissMatch() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader()
+				.getResourceAsStream(JSON_UPDATE_PRACTITIONER_WRONG_ID_PATH)) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java:404]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java [hunk 7]
```diff
@@ -330,4 +346,89 @@ private void verifyUri(String uri) throws Exception {
 		assertThat(results.getEntry().get(0).getResource().getIdElement().getIdPart(), equalTo(PRACTITIONER_UUID));
 	}
 	
+	@Test
+	public void createPractitioner_shouldCreatePractitioner() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_CREATE_PRACTITIONER_PATH)) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java:353]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java [hunk 8]
```diff
@@ -330,4 +346,89 @@ private void verifyUri(String uri) throws Exception {
...
+		        .accept(FhirMediaTypes.JSON).go();
+
+		assertThat(response, isCreated());
+	}
+
+	@Test
+	public void updatePractitioner_shouldUpdateExistingPractitioner() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_PRACTITIONER_PATH )) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java:371]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java [hunk 9]
```diff
@@ -330,4 +346,89 @@ private void verifyUri(String uri) throws Exception {
...
+				.accept(FhirMediaTypes.JSON).go();
+
+		assertThat(response, isOk());
+	}
+
+	@Test
+	public void updatePractitioner_shouldErrorForNoId() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_PRACTITIONER_NO_ID_PATH)) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java:389]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java [hunk 10]
```diff
@@ -330,4 +346,89 @@ private void verifyUri(String uri) throws Exception {
...
+		assertThat(response.getContentAsString(),
+				containsStringIgnoringCase("body must contain an ID element for update"));
+	}
+
+	@Test
+	public void updatePractitioner_shouldErrorForIdMissMatch() throws Exception {
+		String jsonPractitioner;
+		try (InputStream is = this.getClass().getClassLoader()
+				.getResourceAsStream(JSON_UPDATE_PRACTITIONER_WRONG_ID_PATH)) {
+			jsonPractitioner = IOUtils.toString(is);
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java:405]:
```suggestion
			Objects.requireNonNull(is);
			jsonPractitioner = IOUtils.toString(is, StandardCharsets.UTF_8);
```

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java (hunk 1 above)
Comment (gitcliff) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java:68]: i wanted to maintain the status quo as in all other resources

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/PractitionerWebTest_UpdateWithWrongId.json [hunk 11]
```diff
@@ -0,0 +1,4 @@
+{
+  "resourceType": "Practitioner",
+  "id": "a0938123-1691-11df-97a5-7038c432aabc"
+}
```
[the remark this answers is not in the document]
Comment (gitcliff) [reply omod/src/test/resources/org/openmrs/module/fhir2/providers/PractitionerWebTest_UpdateWithWrongId.json]: i have fixed it

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java (hunk 2 above)
Comment (gitcliff) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java]: its done

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java (hunk 3 above)
Comment (gitcliff) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java]: i have adopted it

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java (hunk 4 above)
Comment (gitcliff) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java]: i have adopted it

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java (hunk 5 above)
Comment (gitcliff) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderWebTest.java]: i have adopted it

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java (hunk 8 above)
Comment (gitcliff) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java]: i have adopted it

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java (hunk 9 above)
Comment (gitcliff) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java]: its done

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java (hunk 10 above)
Comment (gitcliff) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderWebTest.java]: i have fixed it

**Comment by ibacher** [comment]
@gitcliff Last thing for this would be to port the tests for the R4 provider back to the R3 provider for the create, update, and delete methods. Thanks for your work on this!

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java [hunk 12]
```diff
@@ -55,6 +59,38 @@ public Practitioner getPractitionerById(@IdParam @NotNull IdType id) {
...
+	
+	@Delete
+	@SuppressWarnings("unused")
+	public org.hl7.fhir.r4.model.OperationOutcome deletePractitioner(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Practitioner practitioner = practitionerService.delete(id.getIdPart());
+		if (practitioner == null) {
+			throw new ResourceNotFoundException("Could not find practitioner to delete with id " + id.getIdPart());
+		}
+		OperationOutcome retVal = new OperationOutcome();
+		retVal.setId(id.getIdPart());
```
Comment (Akayeshmantha) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java:88]: ditto

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProvider.java [hunk 13]
```diff
@@ -57,6 +65,36 @@ public Practitioner getPractitionerById(@IdParam @NotNull IdType id) {
...
+	}
+
+	@Delete
+	@SuppressWarnings("unused")
+	public OperationOutcome deletePractitioner(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Practitioner practitioner = practitionerService.delete(id.getIdPart());
+		if (practitioner == null) {
+			throw new ResourceNotFoundException("Could not find practitioner to delete with id " + id.getIdPart());
+		}
+		OperationOutcome retVal = new OperationOutcome();
```
Comment (Akayeshmantha) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProvider.java:92]: ditto

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java [hunk 14]
```diff
@@ -452,4 +457,64 @@ public void getPractitionerHistoryByWithWrongId_shouldThrowResourceNotFoundExcep
 		assertThat(resourceProvider.getPractitionerHistoryById(idType).isEmpty(), Matchers.is(true));
 		assertThat(resourceProvider.getPractitionerHistoryById(idType).size(), Matchers.equalTo(0));
 	}
+	
+	@Test
+	public void updatePractitioner_shouldUpdatePractitioner() {
+		Practitioner newPractitioner = practitioner;
```
Comment (Akayeshmantha) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java:463]: ditto

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java [hunk 15]
```diff
@@ -452,4 +457,64 @@ public void getPractitionerHistoryByWithWrongId_shouldThrowResourceNotFoundExcep
...
+
+		wrongPractitioner.setId(WRONG_PRACTITIONER_UUID);
+
+		when(practitionerService.update(WRONG_PRACTITIONER_UUID, wrongPractitioner)).thenThrow(MethodNotAllowedException.class);
+
+		resourceProvider.updatePractitioner(new IdType().setValue(WRONG_PRACTITIONER_UUID), wrongPractitioner);
+	}
+
+	@Test
+	public void deletePractitioner_shouldDeletePractitioner() {
```
Comment (Akayeshmantha) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java:492]: ditto

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java [hunk 16]
```diff
@@ -452,4 +457,64 @@ public void getPractitionerHistoryByWithWrongId_shouldThrowResourceNotFoundExcep
...
+		Practitioner newPractitioner = new Practitioner();
+
+		when(practitionerService.delete(PRACTITIONER_UUID)).thenReturn(newPractitioner);
+
+		org.hl7.fhir.r4.model.OperationOutcome result = resourceProvider.deletePractitioner(new IdType().setValue(PRACTITIONER_UUID));
```
Comment (Akayeshmantha) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java:497]: ditto

Comment (Akayeshmantha) [reply api/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderTest.java:497]: ditto ===== previously mentioned :smiley::smiley:

**Comment by gitcliff** [comment]
@ibacher i have added the tests in r3,,,,,,kindly review

File: api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java [hunk 17]
```diff
@@ -60,6 +70,35 @@ public Practitioner getPractitionerById(@IdParam @NotNull IdType id) {
...
+	
+	@Delete
+	@SuppressWarnings("unused")
+	public OperationOutcome deletePractitioner(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Practitioner practitioner = practitionerService.delete(id.getIdPart());
+		if (practitioner == null) {
+			throw new ResourceNotFoundException("Could not find practitioner to delete with id " + id.getIdPart());
+		}
+
+		return OperationOutcome30_40.convertOperationOutcome(FhirProviderUtils.buildDelete(practitioner));
```
Comment (varung-31) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProvider.java:99]: please rewrite this as `FhirProviderUtils.buildDelete(Practitioner30_40.convertPractitioner(practitioner));`

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/PractitionerWebTest_UpdateWithWrongId.json [hunk 18]
```diff
@@ -0,0 +1,8 @@
+{
+  "resourceType": "Practitioner",
+  "id": "a0938123-1691-11df-97a5-7038c432aabc"
+  
+}
```
Comment (varung-31) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/PractitionerWebTest_UpdateWithWrongId.json:5]: Could you please add some more fields to the json files?

**Comment by gitcliff** [comment]
@ibacher kindly review

**Pull request merged**